File: e2e/elixir/test/error_test.exs
# E2e tests for category: error
defmodule E2e.ErrorTest do
  use ExUnit.Case, async: true

  describe "error_401_unauthorized" do
    test "Handles 401 Unauthorized response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_403_forbidden" do
    test "Handles 403 Forbidden response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_404_page" do
    test "Handles 404 response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_408_request_timeout" do
    test "Handles 408 Request Timeout response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_410_gone" do
    test "Handles 410 Gone response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_500_server" do
    test "Handles 500 server error" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_502_bad_gateway" do
    test "Handles 502 Bad Gateway response correctly" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_connection_refused" do
    test "Handles connection refused error gracefully" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_dns_resolution" do
    test "Handles DNS resolution failure gracefully" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_empty_response" do
    test "Handles 200 with completely empty body gracefully" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.html_not_empty) == false
      assert String.trim(result.error.is_error) == false
    end
  end

  describe "error_invalid_proxy" do
    test "Proxy pointing to unreachable address causes connection error during scrape" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_partial_response" do
    test "Handles incomplete or truncated HTTP response" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_rate_limited" do
    test "Handles 429 rate limiting with Retry-After" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_retry_503" do
    test "Retries request on 503 Service Unavailable response" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_retry_backoff" do
    test "Implements exponential backoff when retrying failed requests" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_ssl_invalid_cert" do
    test "Handles SSL certificate validation error" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_timeout" do
    test "Handles request timeout" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_waf_akamai" do
    test "Akamai WAF detection returns WafBlocked error" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_waf_false_403" do
    test "Detects WAF/bot protection false 403 (Cloudflare challenge page)" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end

  describe "error_waf_imperva" do
    test "Imperva/Incapsula WAF detection" do
      assert_raise RuntimeError, fn ->
        Kreuzcrawl.scrape!()
      end
    end
  end
end

File: e2e/rust/tests/error_test.rs
//! E2e tests for category: error

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[tokio::test]
async fn test_error_401_unauthorized() {
    // Handles 401 Unauthorized response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("unauthorized"), "error message mismatch");
}

#[tokio::test]
async fn test_error_403_forbidden() {
    // Handles 403 Forbidden response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("forbidden"), "error message mismatch");
}

#[tokio::test]
async fn test_error_404_page() {
    // Handles 404 response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("not_found"), "error message mismatch");
}

#[tokio::test]
async fn test_error_408_request_timeout() {
    // Handles 408 Request Timeout response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("timeout"), "error message mismatch");
}

#[tokio::test]
async fn test_error_410_gone() {
    // Handles 410 Gone response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("gone"), "error message mismatch");
}

#[tokio::test]
async fn test_error_500_server() {
    // Handles 500 server error
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("server_error"), "error message mismatch");
}

#[tokio::test]
async fn test_error_502_bad_gateway() {
    // Handles 502 Bad Gateway response correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("bad_gateway"), "error message mismatch");
}

#[tokio::test]
async fn test_error_connection_refused() {
    // Handles connection refused error gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("connection"), "error message mismatch");
}

#[tokio::test]
async fn test_error_dns_resolution() {
    // Handles DNS resolution failure gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("dns"), "error message mismatch");
}

#[tokio::test]
async fn test_error_empty_response() {
    // Handles 200 with completely empty body gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let _ = scrape(&engine, &url).await.expect("should succeed");
    // skipped: field 'html_not_empty' not available on result type
    // skipped: field 'error.is_error' not available on result type
}

#[tokio::test]
async fn test_error_invalid_proxy() {
    // Proxy pointing to unreachable address causes connection error during scrape
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("connection"), "error message mismatch");
}

#[tokio::test]
async fn test_error_partial_response() {
    // Handles incomplete or truncated HTTP response
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("data_loss"), "error message mismatch");
}

#[tokio::test]
async fn test_error_rate_limited() {
    // Handles 429 rate limiting with Retry-After
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("rate_limited"), "error message mismatch");
}

#[tokio::test]
async fn test_error_retry_503() {
    // Retries request on 503 Service Unavailable response
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("server_error"), "error message mismatch");
}

#[tokio::test]
async fn test_error_retry_backoff() {
    // Implements exponential backoff when retrying failed requests
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("rate_limited"), "error message mismatch");
}

#[tokio::test]
async fn test_error_ssl_invalid_cert() {
    // Handles SSL certificate validation error
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("ssl"), "error message mismatch");
}

#[tokio::test]
async fn test_error_timeout() {
    // Handles request timeout
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("timeout"), "error message mismatch");
}

#[tokio::test]
async fn test_error_waf_akamai() {
    // Akamai WAF detection returns WafBlocked error
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    // skipped: field 'error.is_waf_blocked' not available on result type
}

#[tokio::test]
async fn test_error_waf_false_403() {
    // Detects WAF/bot protection false 403 (Cloudflare challenge page)
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    assert!(result.is_err(), "expected call to fail");
    assert!(result.as_ref().unwrap_err().to_string().contains("forbidden"), "error message mismatch");
    // skipped: field 'error.is_waf_blocked' not available on result type
}

#[tokio::test]
async fn test_error_waf_imperva() {
    // Imperva/Incapsula WAF detection
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await;
    assert!(result.is_err(), "expected call to fail");
    // skipped: field 'error.is_waf_blocked' not available on result type
}
